### PR TITLE
#40 Add jsonMapper into mappers

### LINKED ISSUE #40: WIP: Add jsonMapper into mappers
Part of #21 

# TODO:

1. [ ] Create file `src/mappers/jsonMapper.ts` with generic function matches to type **EDIT:** ~`ValueMapper<TJsonResult>`~ `ValueMapper<any>` . Which parse json string into `TJsonResult` and return default `null` when failure.
2. [ ] reexport it in `src/mappers/index.ts` 
3. [ ] Prove that it works in unit test
4. [ ] Describe it in Readme.md

## proposed test data provider:

```js
const dataProvider = [
  { inValue: '', expectedResult: null },
  { inValue: null, expectedResult: null },
  { inValue: ' ', expectedResult: null },
  { inValue: 'asd', expectedResult: null },
  { inValue: 'null', expectedResult: null },
  { inValue: 'false', expectedResult: false },
  { inValue: 'true', expectedResult: true },
  { inValue: '0', expectedResult: 0},
  { inValue: '1', expectedResult: 1},
  { inValue: '"string"', expectedResult: 'string'},
  { inValue: '\'string\'', expectedResult: null},
  { inValue: '{"a":1}', expectedResult: {a:1}},
  { inValue: '{"a":[1]}', expectedResult: {a:[1]}},
]
```

For making a reservation please open draft PR with this issue linked

### PR DESCRIPTION
**Pull request by Gontrum**
If `TJsonResult` should be put somewhere else, just let me know :-) 
Should resolve #40 

- added some tests for jsonMapper
- added jsonMapper to README
- added jsonMapper and a TJsonResponse-Type

### CODE REVIEW

File: src/mappers/jsonMapper.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    }
+}
+
+interface IJsonResult {
+    [key: string]: string | boolean | number
+}
+
+type JsonMapper = ValueMapper<IJsonResult | null>
+
+export const jsonMapper: JsonMapper = parseJSON
```
Comment (Siemienik) [inline src/mappers/jsonMapper.ts:17]: I consider about one of two way to implement jsonMapper. Please choose this one which you prefer:

1. The simplest, enough for in `xlsx-import` usages
```ts
export const jsonMapper: ValueMapper<any> = (value: string): any => {
    try {
        return JSON.parse(value)
    } catch (e) {
        return null
    }
}
```

2. More complex, and better typed way:

```ts
type JsonMapper= <TJsonResult>(value: string) => TJsonResult|null;

export const jsonMapper:JsonMapper = <TJsonResult extends any>(value: string): TJsonResult|null => {
    try {
        return JSON.parse(value)
    } catch (e) {
        return null
    }
}```

**Review by Siemienik — CHANGES_REQUESTED** [review]
Thank you for contribution ❤️ 
I have one proposition for you described in a comment.

File: src/mappers/jsonMapper.ts (hunk 1 above)
Comment (Gontrum) [reply src/mappers/jsonMapper.ts]: Hi @Siemienik,

thank you for your great support! I understood it now 🙂

File: samples/nodejs-ts/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# NodeJS + Typescript
```
Comment (Siemienik) [inline samples/nodejs-ts/README.md:1]: I don't know why, but in your PR appeared changes from other. Probably rebasing in 6730388598f8aec243213284b21f517db5248af3 went wrong.

Could I please you to reset code into 77d01f3a5c22c5f9964f05ca7656b9e1da2dbefe ?

```
git reset --hard 77d01f3a5c22c5f9964f05ca7656b9e1da2dbefe
```
and later use `git merge` commend to get changes from upstream master ?

**Review by Siemienik — COMMENTED** [review]
I has only one comment about rebase failed, everything else looks good to me 👍

**Review by Siemienik — APPROVED** [review]
I did merge instead of rebase. 
LGTM now 😄 

Thank you for this contribution, great job 👍

**Pull request merged**